### DISCUSSION TITLE
Is there a recommended solution on how to enable mongoDB "change streams" ?

### DISCUSSION THREAD

**Discussion by bodote** [flapdoodle-oss/de.flapdoodle.embed.mongo#496] [Q&A] [answered]
MongoDB has a feature called ["change streams"](https://www.mongodb.com/docs/v7.0/changeStreams). However, this feature is only available if there is a "replica set". The replica set in turn is NOT enabled by default for single instance MongoDBs. You can enable the replica set with  de.flapdoodle.embed.mongo. But it seemes a little bit cumbersome though.

Is there any recommendation on how to use this with de.flapdoodle.embed.mongo ?

what we do is this: 

```
 @Override
            public Transition<MongodArguments> mongodArguments() {
                return Start.to(MongodArguments.class)
                        .initializedWith(MongodArguments.defaults()
                                .withReplication(Storage.of("testRepSet", 5000))
                                .withUseNoJournal(false));
            }
```
plus : 
```
 mongoClient.getDatabase("admin").runCommand(new Document("replSetInitiate", new Document()));
```
immediately `@BeforeEach` test.

especially the need for this admin - command above turned out to be somehow problematic and lead to unstable test until we wraped it part of the test code in  some  `await().atMost(10, SECONDS).until(() -> {...}` 

Is there any better way ?

**Comment by michaelmosmann** [comment]
@bodote i will have a look at this.. do you have any error message for the "unstable" case?

**Reply by bodote** [reply]
I think that this is a timeing / race condition problem, e.g. when i try to access the mongoDB immediately after sending the 
`mongoClient.getDatabase("admin").runCommand(new Document("replSetInitiate", new Document()));`
then sometimes i got the error that the database is not in primary or secondary state. When i add a Thread.sleep(500) the problem disapears.

**Reply by bodote** [reply]
and it would by nicer if the de.flapdoodle.embed.mongo library would itself take care of the 
`mongoClient.getDatabase("admin").runCommand(new Document("replSetInitiate", new Document()));`
command , wouldn't it?

**Reply by michaelmosmann** [reply]
@bodote I am not sure about that:) as there are unlimited ways to work with a mongodb database and i can not handle all of them.. but maybe it is something very simple.. as some kind of listener which is triggered if mongodb changes its state..

**Reply by michaelmosmann** [reply]
as @XSpielinbox noticed in the issue.. if you are using the spring integration i did write some code to handle this replicaset stuff: https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo.spring/blob/041421695626591599798602a39df8a61e3fdb42/src/main/java/de/flapdoodle/embed/mongo/spring/autoconfigure/AbstractServerFactory.java#L84

maybe i should backport it to the mongodb project

**Comment by michaelmosmann** [comment, marked as answer]
@bodote i will create an issue from that..

**Reply by michaelmosmann** [reply]
--> https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/issues/497

**Comment by michaelmosmann** [comment]
@bodote .. i close this. Reopen if needed..
